backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/TagService.java

```diff
@@ -0,0 +1,105 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.hyperledger.bpa.impl;
+
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.data.exceptions.DataAccessException;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.hyperledger.bpa.api.TagAPI;
+import org.hyperledger.bpa.api.exception.WrongApiUsageException;
+import org.hyperledger.bpa.config.TagConfig;
+import org.hyperledger.bpa.model.Tag;
+import org.hyperledger.bpa.repository.TagRepository;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.*;
+
+@Slf4j
+@Singleton
+public class TagService {
+
+    @Inject
+    TagRepository tagRepo;
+
+    @Inject
+    TagConfig configuredTags;
+
+    public @Nullable TagAPI addTag(@NonNull String name) {
+        return addTag(name, Boolean.FALSE);
+    }
+
+    TagAPI addTag(@NonNull String name, @NonNull Boolean isReadOnly) {
+        Tag tag = Tag.builder()
+                .name(name)
+                .isReadOnly(isReadOnly)
+                .build();
+
+        Tag saved;
+        try {
+            saved = tagRepo.save(tag);
+        } catch (DataAccessException e) {
+            throw new WrongApiUsageException("Tag with name: " + name + " already exists.");
+        }
+
+        return TagAPI.from(saved);
+    }
+
+    public Optional<TagAPI> updateTag(@NonNull UUID id, @NonNull String name) {
+        Optional<Tag> dbTag = tagRepo.findById(id);
+        if (dbTag.isPresent()) {
+            tagRepo.updateNameById(dbTag.get().getId(), name);
+            log.debug("Updating existing tag name {} with new name: {}", dbTag.get(), name);
+            dbTag.get().setName(name);
+            return Optional.of(TagAPI.from(dbTag.get()));
+        }
+        return Optional.empty();
+    }
+
+    public List<TagAPI> listTags() {
+        List<TagAPI> result = new ArrayList<>();
+        tagRepo.findAll().forEach(dbT -> result.add(TagAPI.from(dbT)));
+        return result;
+    }
+
+    public Optional<TagAPI> getTag(@NonNull UUID id) {
+        Optional<Tag> tag = tagRepo.findById(id);
+        return tag.map(TagAPI::from);
+    }
+
+    public void deleteTag(@NonNull UUID id) {
+        tagRepo.findById(id).ifPresent(s -> tagRepo.deleteById(id));
+    }
+
+    public void resetWriteOnlyTags() {
+        long deleted = tagRepo.deleteByIsReadOnly();
+        log.debug("Removed {} preconfigured tags", deleted);
+        if (configuredTags.getTags() != null) {
+            configuredTags.getTags().forEach(t -> {
+                try {
+                    TagAPI tagAPI = addTag(t, Boolean.TRUE);
+                    log.debug("Added preconfigured tag with name: {}", tagAPI.getName());
+                } catch (DataAccessException e) {
+                    log.warn("Tag with name {} will not be added", t, e);
+                }
+            });
+        }
+
+    }
```

Comment:
This strategy taken from schemas won't work unfortunately. As soon as a tag is given to partner it can't be simply deleted and reinitialized from config.